### PR TITLE
Use Docker in .travis.yml

### PR DESCRIPTION
**Pull request by DoomHammer**
First, slightly hackish approach. It should handle bottling for shorter builds at least.

### CODE REVIEW

**Comment by dpo** [comment]
Is there a way to package Java in the docker images? Look at this "successful" build: https://travis-ci.org/Homebrew/homebrew-science/jobs/86294570.

**Comment by sjackman** [comment]
Sure, Java can be installed with apt-get or brew. It'll have to wait until I'm back from vacation in the new year.

**Comment by DoomHammer** [comment]
With all the talk about leaving Homebrew organization maybe it is a good time to revisit those changes as Linuxbrew would definitely benefit from any kind of working CI?

**Comment by sjackman** [comment]
Yes, absolutely. Getting Travis CI up and running is my next project after Linuxbrew moves to its own organization.

**Comment by DoomHammer** [comment]
I'll wait here, then :watch:

**Comment by sjackman** [comment]
Travis CI for Linuxbrew is up and running! I haven't looked into Docker yet.

**Comment by DoomHammer** [comment]
Let me know when you want to look at it, I'll rebase the changes then.

**Comment by sjackman** [comment]
I had planned on building bottles on the `sjackman/linuxbrew-standalone` Docker image. I'm starting to think that there are some benefits to building the bottles on an older distribution. The compiled bottles should work on Linuxbrew Standalone, but should also work on older systems. The dependencies used from the host system should be `glibc` and `libstdc++`. All the other dependencies should be provided by Linuxbrew. That way perhaps Linuxbrew users could use bottles without having to install `glibc` and `gcc` (for `libstdc++`).

Linuxbrew: glibc 2.19

Ubuntu 14: glibc 2.19
Ubuntu 12: glibc 2.15

CentOS 7: glibc 2.17
CentOS 6: glibc 2.12
CentOS 5: glibc 2.5

**Comment by DoomHammer** [comment]
So, do you suggest using Ubuntu 12 as a base for Docker/Travis instead of current `linuxbrew-standalone`?

**Comment by sjackman** [comment]
I've been using `Ubuntu 14` as the base for Docker/Travis instead of `linxubrew-standalone`. It has the same version of `glibc` used by Linuxbrew (2.19), which is a nice coincidence. The bottles built on Ubuntu 14 work great on Linuxbrew. I've tested a few now.

**Comment by sjackman** [comment]
Hi, Piotr. I've created a new [Dockerfile](https://github.com/Linuxbrew/linuxbrew/blob/master/Dockerfile) for Linuxbrew, and there's a new [Linuxbrew/linuxbrew](https://hub.docker.com/r/linuxbrew/linuxbrew/) docker image created from it.

**Comment by sjackman** [comment]
`brew tests` is oddly failing right now.
```
1) Failure:
IntegrationCommandTests#test_uses [/home/linuxbrew/.linuxbrew/Library/Homebrew/test/test_integration_cmds.rb:629]:
--- expected
+++ actual
@@ -1,2 +1,2 @@
-"testball1
-testball2"
+"testball2
+testball1"
2) Failure:
LanguageModuleRequirementTests#test_good_python_deps [/home/linuxbrew/.linuxbrew/Library/Homebrew/test/test_language_module_requirement.rb:43]:
Expected #<LanguageModuleRequirement: "languagemodule" [:python, "datetime", nil]>
to be satisfied?.
```

**Comment by DoomHammer** [comment]
I have uploaded a new version. I guess all those `if [ "${TRAVIS_OS_NAME}" = "osx" ]` are no longer necessary? Or do you plan to support OS X in the long term?

**Comment by DoomHammer** [comment]
Also: I am not sure whether test-bot behaves correctly, since it tries to tap `doomhammer/linuxbrew` which in practice means `doomhammer/homebrew-linuxbrew` (which does not exist).

**Comment by sjackman** [comment]
> Or do you plan to support OS X in the long term?

For the package manager `brew` and for for the formula, but not for the `.travis.yml`.

**Comment by sjackman** [comment]
> Also: I am not sure whether test-bot behaves correctly, since it tries to tap doomhammer/linuxbrew which in practice means doomhammer/homebrew-linuxbrew (which does not exist).

That's a bug. Feel free to submit a pull request if you are able to fix it.

**Comment by DoomHammer** [comment]
I am not familiar with `brew` internals. Could you point me what is the best way to start, since `test-bot.rb` is pretty big?

**Comment by DoomHammer** [comment]
I believe I have found the problematic line, but I don't know the original authors idea. I believe @xu-cheng have added it.

Why is there `if user == "Homebrew" && repo == "homebrew"`? Doesn't it prevent Pull Requests from being tested (as they clearly have a different user, but the same repo name)?

File: Library/Homebrew/tap.rb [hunk 1]
```diff
@@ -31,7 +31,7 @@ def self.fetch(*args)
     repo = repo.strip_prefix "homebrew-"
 
     if user == "Homebrew" && repo == "homebrew" ||
-        user == "Linuxbrew" && repo == "linuxbrew"
+        repo == "linuxbrew"
```
Comment (xu-cheng) [inline Library/Homebrew/tap.rb]: This is incorrect. The core tap should always call `Homebrew/homebrew` even for Linuxbrew. Noted this is not the name of Linuxbrew project but  an internal name for a specialize tap.

Comment (DoomHammer) [inline Library/Homebrew/tap.rb]: Yes, I believe this is incorrect, I only wanted to test it. In my case I get `user == "DoomHammer" && `repo == "linuxbrew"` when running `test-bot` on my PR on Travis. Could you help me find out what is wrong here?

Comment (xu-cheng) [inline Library/Homebrew/tap.rb]: You need to update https://github.com/Linuxbrew/linuxbrew/blob/master/Library/Homebrew/cmd/test-bot.rb#L44
to
```ruby
if (tap = ENV["TRAVIS_REPO_SLUG"]) && (tap =~ %r{^([\w-]+)/(homebrew-)?([\w-]+)$})
```

Comment (xu-cheng) [inline Library/Homebrew/tap.rb]: Sorry, the regex should be `%r{^([\w-]+)/homebrew-([\w-]+)$}`. I wrote it wrong in the first time.

**Comment by DoomHammer** [comment]
After dealing with missing dependency on `rake` (by running `brew install ruby`) I have issues with `brew tests` as seen in [https://travis-ci.org/DoomHammer/linuxbrew/builds/114097361](Travis log). Any ideas what am I missing?

Most interestging lines are:

```
 1) Failure:
FormulaTests#test_eligible_kegs_for_cleanup [/home/linuxbrew/.linuxbrew/Library/Homebrew/test/test_formula.rb:365]:
--- expected
+++ actual
@@ -1,2 +1,2 @@
 [#<Keg:/tmp/homebrew_tests20160306-1572-kmxsq4/cellar/testball/0.2>,
- #<Keg:/tmp/homebrew_tests20160306-1572-kmxsq4/cellar/testball/0.3>]
+ #<Keg:/tmp/homebrew_tests20160306-1572-kmxsq4/cellar/testball/0.1>]
```

**Comment by sjackman** [comment]
That's a weird failure. I'm not sure why the test is saying that 0.2 and 0.3 are expected to be cleaned up. Perhaps you could troubleshoot by printing out the value of `f3.installed_kegs` right before:
```ruby
    assert_equal f3.installed_kegs[0..1], f3.eligible_kegs_for_cleanup
```

**Comment by DoomHammer** [comment]
Running this locally I get slightly different error, but:

```
KEGGIE: [#<Keg:/tmp/homebrew_tests20160307-182-ykgdsq/cellar/testball/0.3>, #<Keg:/tmp/homebrew_tests20160307-182-ykgdsq/cellar/testball/0.1>, #<Keg:/tmp/homebrew_tests20160307-1
82-ykgdsq/cellar/testball/0.2>]
```

then

```
  1) Failure:
FormulaTests#test_eligible_kegs_for_cleanup [/home/linuxbrew/.linuxbrew/Library/Homebrew/test/test_formula.rb:366]:
--- expected
+++ actual
@@ -1,2 +1,2 @@
-[#<Keg:/tmp/homebrew_tests20160307-182-ykgdsq/cellar/testball/0.3>,
- #<Keg:/tmp/homebrew_tests20160307-182-ykgdsq/cellar/testball/0.1>]
+[#<Keg:/tmp/homebrew_tests20160307-182-ykgdsq/cellar/testball/0.1>,
+ #<Keg:/tmp/homebrew_tests20160307-182-ykgdsq/cellar/testball/0.2>]
```

Also:

```
  2) Failure:
IntegrationCommandTests#test_uses [/home/linuxbrew/.linuxbrew/Library/Homebrew/test/test_integration_cmds.rb:629]:
--- expected
+++ actual
@@ -1,2 +1,2 @@
-"testball1
-testball2"
+"testball2
+testball1"
```

**Comment by sjackman** [comment]
I've seen this test failure too. See #895

**Comment by DoomHammer** [comment]
Now the #895 is gone, Travis CI is passing. I understand it is better to wait for [upstream](https://github.com/Homebrew/homebrew/commit/7970035e5d8bdaf201e44b0fa84a047f9f67ab22) to be merged rather than committing local changes? BTW: What is this ci/dockercloud thing?

**Comment by sjackman** [comment]
I've merged Homebrew into Linuxbrew just now. I'm testing out using Docker Cloud running on Amazon EC2 as a CI for building bottles for Linuxbrew.
https://docs.docker.com/docker-cloud/feature-reference/automated-build/

**Comment by DoomHammer** [comment]
Well, when I click "Details" I get 404 and I wanted to know what did I break (as it says "Your tests failed in Docker Cloud").

**Comment by sjackman** [comment]
The link works for me. I'll guess that it's a permissions issue, that you don't have the perms to view the build log. Once it's up and running I'll sort that out, hopefully. Here's the log:
```
==> brew doctor
==> FAILED
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry and just ignore them. Thanks!
Warning: Suspicious git origin remote found.
With a non-standard origin, Homebrew won't pull updates from
the main repository. The current git origin is:
https://9a1bc74ab023883d39a8ba7b145913ebb40de5e1@github.com/DoomHammer/linuxbrew.git
Unless you have compelling reasons, consider setting the
origin remote to point at the main repository, located at:
https://github.com/Linuxbrew/linuxbrew.git
```
The error message is because `brew test-bot` isn't yet set up to integrate with Docker Cloud CI. I'm working on it.

**Comment by sjackman** [comment]
What about the Travis failure though. Does that make sense to you?

**Comment by DoomHammer** [comment]
Ah, so I plead not guilty, then for Docker CI. As for the Travis, it just timed out. I made it re-run the tests, but they are at the edge of timing-out too. Building `ruby` is the most time-consuming thing here, but couldn't do it properly without it. One possible idea would be to have Docker image prebuilt with `ruby`.

**Comment by DoomHammer** [comment]
The little guy is green now :)

**Comment by DoomHammer** [comment]
Sorry to be bothering you, @sjackman, but are we go with this?

**Comment by sjackman** [comment]
> Building ruby is the most time-consuming thing here

Can we use the Ruby deb rather than brew ruby?

**Comment by sjackman** [comment]
No apologies needed for bothering me. What are the advantages as you see it of using a Docker container rather than using the usual Travis build bot?

**Comment by DoomHammer** [comment]
Ruby itself is not a problem. `brew` uses `rake` which is a separate package on Debian and somehow it didn't seem to be doing alright.

As for why Docker? Honestly, can't remember but this PR was a result of some discussion, IIRC. History says it was probably #528 .

One thing for certain: playing with Docker is more reproducible, so theoretically anybody can look at Travis error and then check it locally.

**Comment by sjackman** [comment]
> One thing for certain: playing with Docker is more reproducible, so theoretically anybody can look at Travis error and then check it locally.

That is in advantage. It is quite frustrating trying to reproduce a Travis build failure locally. Do you know whether the Travis build machine images are available on Docker Hub?

**Comment by DoomHammer** [comment]
No idea about the Travis build machines but in my PR I am making use of `linuxbrew/linuxbrew` (same thing I usually do prior to reporting bug, as this image I consider 'sane' contrary to most of machines I work on).

**Comment by sjackman** [comment]
Hi, Piotr. I would like to continue testing Ubuntu 12 precise using Travis CI. It's important to test that installing `glibc` and `gcc` automatically works for older distributions.
Ubuntu 14 trusty is being tested using Docker cloud. That's my biggest reason for not using this patch. I also want to test Linuxbrew on a variety of platforms, and `linuxbrew/linuxbrew` is already being tested on Docker Cloud.
I know that this patch precedes my use of Docker Cloud, which is relatively recent.
I have thought about enabling Travis testing on both Ubuntu 12 and Ubuntu 14, but not necessarily using Docker. The downside to adding more dists to the testing matrix, is that tests will take longer, and it's nice to get quick feedback when iteratively testing updating a PR.

**Comment by DoomHammer** [comment]
Makes sense to me. Closing this then?

**Comment by sjackman** [comment]
Yes, I think so. Sorry, Piotr. This PR raised some helpful discussion. Thanks for your effort.

**Comment by DoomHammer** [comment]
No problem. Not all ideas are good ones ;)

**Pull request closed without merging**